File: week-06/readability.py
# TODO
import re
# A program that first asks the user to type in some text, and then outputs the grade level for the text, according to the Coleman-Liau formula
# Coleman-Liau index is computed as 0.0588 * L - 0.296 * S - 15.8, where L is the average number of letters per 100 words in the text, and S is the average number of sentences per 100 words in the text

# Write Coleman-Liau score index function


def get_coleman_liau_index(avg_letters, avg_sentences):
    return round(0.0588 * avg_letters - 0.296 * avg_sentences - 15.8)


# Get inptut and convert it to lower case
sentence = input("Text: ").lower()

# Count letters, words and sentences
# For letters I'll do regex pattern matching only lowercase letters
letters = len(re.findall(r"[a-z]", sentence))
words = len(sentence.split(" "))
# For sentences I'll do regex pattern matching for ending symbols
sentences = len(re.split(r"[.!?]", sentence))-1

# Calculate averages of letters per 100 words and sentences per 100 words
avg_letters = letters * 100 / words
avg_sentences = sentences * 100 / words

# Run processed user input through Coleman-Liau function and get results
coleman_liau_score = get_coleman_liau_index(avg_letters, avg_sentences)

# Print sentence result based on assignment rules
if (coleman_liau_score < 1):
    print("Before Grade 1")
elif (coleman_liau_score > 16):
    print("Grade 16+")
else:
    print(f"Grade {coleman_liau_score}")

File: week-02/readability.c
#include <cs50.h>
#include <stdio.h>
#include <string.h>
#include <ctype.h>
#include<math.h>

int get_coleman_liau_result(float l, float s);
int count_letters(string text);
int count_words(string text);
int count_sentences(string text);

int main(void)
{

    // Get text from user
    string text = get_string("Text: ");

    // Count sentences, words and letters
    int letters = count_letters(text);
    int words = count_words(text);
    int sentences = count_sentences(text);

    // Get average number of letters per 100 words
    float letters_in_100w = ((float)(letters)) / words * 100;

    // Get averate number of sentences per 100 words
    float sentences_in_100w = ((float)(sentences)) / words * 100;

    // Get Coleman-Liau score index
    int grade = get_coleman_liau_result(letters_in_100w, sentences_in_100w);
    printf("%i\n", grade);

    // Output result Grade x
    if (grade < 1)
    {
        printf("Before Grade 1\n");
    }
    else if (grade > 16)
    {
        printf("Grade 16+\n");
    }
    else
    {
        printf("Grade %i\n", grade);
    }
}


int get_coleman_liau_result(float l, float s)
{
    float result = 0.0588 * l - 0.296 * s - 15.8;
    int r = round(result);
    return r;
}

int count_letters(string text)
{
    int count = 0;
    for (int i = 0; i < strlen(text); i++)
    {
        if (tolower(text[i]) >= 'a' && tolower(text[i]) <= 'z')
        {
            count++;
        }
    }
    return count;
}

int count_words(string text)
{
    int count = 1;
    for (int i = 0; i < strlen(text); i++)
    {
        // typecasting
        int l = (int)(text[i]);
        // If char is 32 (asci for space) and one before character was not 32 (so we can have multiple spaces)
        if (l == 32 && ((int)(text[i - 1]) != 32))
        {
            count++;
        }
    }
    return count;
}

int count_sentences(string text)
{
    int count = 0;
    for (int i = 0; i < strlen(text); i++)
    {
        // typecasting
        int l = (int)(text[i]);
        // If char is !, . or ?
        if (l == 33 || l == 46 || l == 63)
        {
            count++;
        }
    }
    return count;
}